Nodal interpolation for NURBS

IdoAkkerman opened the pull request:
Changed PR header (27 June 2024)
This PR adds a Projector to the NURBS-based elements.
Furthermore, initialization of the a NURBS-based` FESpace` made to look similar to the initialization of the `FESpace` for other FEs.
By performing the selection of the `FE_Collection` in  small utility functions, we can run NURBS-based discretisations in for some of the examples with minimal changes.
Some small changes were made to some NURBS classes, `KnotVector`  most notably. Some examples where slightly modified to accommodate this.

Notes on the  `Project` member function of the different NURBS-based FE:
The `coefficient` is evaluated at the Bottella points,  this value is directly injected in the solution vector for the shape function associated with the Botella point. 
The Botella point is the point where the shapefunction attains its maximum.
NOTE: this is not a true projector in the mathematical sense, applying the 
operation a second time might change the solution.


ORIGINAL MESSAGE (23 May 2024)
By inheriting  from `NodalFiniteElement` instead of `ScalarFiniteElement` (the base class of `NodalFiniteElement` ), we get a free _nodal_ interpolator.

This still  needs some fundamental and practical checking.

Please note:

That the terminology of _projection_ is not correct in this context instead it is a interpolation(I think??).
The idempotent property is not satisfied, a "projection of a projection" should be the same as the original projection.

I do not expect this to be a fundamental problem, I guess  an interpolation would also work in most (all?) situations, or am I wrong on this?

Changing **all** the projection calls in interpolation calls would be way to excessive (but technically correct IMO).
<!--GHEX{"author":"IdoAkkerman","assignment":"2024-07-14T18:02:59","editor":"tzanio","id":4326,"reviewers":["v-dobrev","justinlaughlin","tzanio","dylan-copeland"],"merge":"2026-01-21T16:53:07","approval":"2026-01-21T16:48:16"}XEHG-->
<!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4326](https://github.com/mfem/mfem/pull/4326) | @IdoAkkerman | @tzanio | @v-dobrev + @justinlaughlin + @tzanio + @dylan-copeland | 7/14/24 | 1/21/26 | 1/21/26 | |
<!--ELBATXEHG-->
<details>
<summary>PR Checklist</summary>
    
- [x] Code builds.
- [x] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [x] Update `INSTALL`:
    - [x] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [x] Have the version ranges for any required or optional libraries changed?
    - [x] Does `make` or `cmake` have a new target?
    - [x] Did the requirements or the installation process change? *(rare)*
- [x] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [x] `.github`
    - [x] `.appveyor.yml`
- [x] Update `.gitignore`:
    - [x] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [x] Add new patterns (just for the new files above) and re-run the above test.
- [x] New examples:
    - [x] All sample runs at the top of the example source file work.
    - [x] Update `examples/makefile`:
      - [x] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [x] Add any files generated by it to the `clean` target.
      - [x] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [x] Update `examples/CMakeLists.txt`:
      - [x] Add the example code to the `ALL_EXE_SRCS` variable.
      - [x] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [x] List the new example in `doc/CodeDocumentation.dox`.
   - [x] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [x] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [x] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [x] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [x] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [x] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [x] New miniapps:
   - [x] All sample runs at the top of the miniapp source file work.
   - [x] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [x] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [x] Update CMake build system:
      - [x] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [x] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [x] Consider adding a new test for the new miniapp.
   - [x] List the new miniapp in `doc/CodeDocumentation.dox`
   - [x] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [x] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [x] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [x] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [x] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [x] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [x] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [x] Consider adding new sample runs in existing examples to highlight the new capability.
   - [x] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [x] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [x] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [x] Update this checklist, if the new pull request affects it.
- [x] Run `make unittest` to make sure all unit tests pass.
- [x] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

v-dobrev commented:
`NodalFiniteElement`s are elements such that the DOF functionals, $\\{ f_i \\}$, and the basis functions, $\\{ \phi_i \\}$, satisfy the property: $f_i(\phi_j) = \delta_{ij}$, locally on each element. This is not true for the NURBS elements since such $\\{ f_i \\}$ cannot be defined locally and require a global solve. Or, am I missing something?

v-dobrev commented:
For `PositiveFiniteElement`s, we define the `Project` methods as to compute the local DOFs as the values at the local "nodes" which are simply defined to be a uniform set of points in the reference space including points on the boundary of the element. This, of course, is not a real projection, just an approximation which has the important property that it preserves positivity.

In principle, such "projection" is well defined for NURBS spaces, however, it is not local -- the "nodes" associated with some of the basis functions that have support in a given mesh element lie in a different mesh element. For this case, what should work best is to NOT use the local `Project` methods but implement the interpolation directly in the global methods like `GridFunction::ProjectCoefficient`.

v-dobrev commented:
The NURBS version of `ProjectCoefficient` can be implemented patch-wise with some help from the knot vectors -- they should tell us in which 1D-element the "node" for a given 1D basis function lies, and what is the "node" location within that 1D element.

IdoAkkerman commented:
@v-dobrev You are correct $f_i(\phi_j) = \delta_{ij}$ is not satisfied for NURBS.
However, the routine that determines the interpolation (and project) if this kronecker-delta property holds, still provides a reasonable **approximation** in case of NURBS (it is neither an interpolation or a projection I now realize), I believe.

If the approximate function is written as $g^h(\phi) = \sum a_i f_i(\phi)$ where  $a_i = g(\phi_i)$
then $g^h(\phi_j) = g(\phi_j) $ if $f_i(\phi_j) = \delta_{ij}$. 

For NURBS this will not be true.
However, I do believe $g^h$ is a reasonable approximation $g^h(\phi) \approx g(\phi)$, similar to "Bernstein polynomial approximation" as mentioned Farouki (2012). This probable used the same intuition (or formal proof?) as the `PositiveFiniteElement` approach you mention.

I have implemented NURBS specific routine  in `GridFunction::ProjectCoefficient`, which does a $H_1$ projection and assumes a serial implementation (this is in the #3999).   This does not preserve positivity.

I do not understand the problems with `Project`... perhaps I need to think harder to formulate a meaningful question.
I have tried the switch to `NodalFiniteElement` and it seems to work, and preserve positivity.
My guess is that is similar to `PositiveFiniteElement` but with an implicit choice for the points.

Note: the coefficient I use was generated using a function. So the (Trans,ip)  pair gets converted to a actual coordinate, which does not give issues if it is outside the element (in the generic case this might be an issue. )

IdoAkkerman commented:
@v-dobrev I am starting to realize what the issues are... Long story short.. the routine does not exactly what hoped (thoughed) is would do. The abstract approach of MFEM does not allow my short cut in general.

IdoAkkerman commented:
I am now implementing a nodal evaluation at Greville / Demo(?) points. 
I have two very different questions:

1) For each shape there is an associated point. The  value of the interpolated function at this point is directly injected 
as the coefficient for the associated shape.
The points associated to the shapes defined on an element can be outside the current element, as already alluded to by @v-dobrev.  Currently this works... but this is likely due to the `Coefficient` being a `FunctionCoefficient`, and because the meshes are nicely uniform with affine mappings.
a) I think it would be beter to avoid this _outside-the-element_ evaluation. Or am I to strict?
b) What is the best way to avoid this _outside-the-element_ evaluation. I see 2 options.
Op1) Only evaluated the dofs that are  _inside-the-element_  and provide a mechanism to indicate which dofs are evaluated.
This could be done by assign a very large number (or NAN or other marker) to the dof, and modify the `SetVector` routine to handle this.
Alternatively, this can be done by adding the dofindices to the argument list and reduce it in the routine to the dofs that are actually evaluated.  

Op2) A dedicated ProjectCoeffficient routine in the GridFunction is possible, this would be simpler initially. But implicit ifs dut to polymorphism will be replaced by explicit ifs, which at the end will create a difficult mess (I think).



2) Does anybody have a clou how the Demko points are defined and or computed? The reference I have looked at (Original Demko paper from 85 and C.de Boor Practical Spline book) are not helping...
What I have implemented is the Maximum of the Shape functions. I wonder if this is the same as the Demko points?

IdoAkkerman commented:
I have implemented the Demko points using the "Remez" algorithm.

IdoAkkerman commented:
I made some small corrections and renaming in the `KnotVector` class.
I also added some small features to the `KnotVector` class.

I renamed `findSpanLocation` to `GetKnot`, and  `findKnotSpan` to `GetSpan`.
NOTE: that `GetSpan` not has an explicit check on the front **and** back of the vector, and the strange shift by 1 (inherited from fortran) has been corrected.

I also added the Greville, Botella and Demko Points to the `KnotVector` class.
I added some unittests to test the new routines.

I also added a `Project` member to the `NURBS?DFiniteElements`. These only evaluate dofs that are in the element.
Otherwise a very high value is set.
In `GridFunction` this is detected and corrected for. This is placeholder stuff, and needs to be modified to an appropriate mechanism.

IdoAkkerman commented:
I think this PR is ready for review. I edited the first comment in the thread to say:

This PR adds a Projector to the NURBS-based elements.
Initialization of the a NURBS-based FESpace made to look similar to the initialization of the FESpace for other FEs.
By performing the selection of the FE_Collection in small utility functions, we can run NURBS-based discretisations in for some of the examples with minimal changes.

Notes on the Project member function of the different NURBS-based FE:
The coefficient is evaluated at the Bottella points, this value is directly injected in the solution vector for the shape function associated with the Botella point.
The Botella point is the point where the shapefunction attains its maximum.
NOTE: this is not a true projector in the mathematical sense, applying the
operation a second time might change the solution.

dylan-copeland commented on mesh/nurbs.cpp:
```diff
@@ -574,6 +829,37 @@ void KnotVector::FindInterpolant(Array<Vector*> &x)
+
+   // Assemble collocation matrix
+   Vector shape(Order+1);
+   DenseMatrix A(NumOfControlPoints,NumOfControlPoints);
```
For a square matrix, the size can be given just once.

IdoAkkerman replied:
Simplified.

dylan-copeland commented on mesh/nurbs.cpp:
```diff
@@ -574,6 +829,37 @@ void KnotVector::FindInterpolant(Array<Vector*> &x)
+void KnotVector::GetInterpolant(const Vector &x, const Vector &u,
+                                Vector &a) const
+{
+   int ks;
```
Declare `ks` and `xi` as `const` inside the loop.

dylan-copeland commented on mesh/nurbs.cpp:
```diff
@@ -574,6 +829,37 @@ void KnotVector::FindInterpolant(Array<Vector*> &x)
+      }
+   }
+
+   // Solve problems
```
Comment that this matrix is banded, and a banded solver could be used in the future for optimization?

dylan-copeland commented on fem/fe_coll.cpp:
```diff
@@ -394,6 +394,50 @@ FiniteElementCollection *FiniteElementCollection::New(const char *name)
+{
+   if (NURBS)
+   {
+      //return new NURBSFECollection(order);
```
Not supported yet?

IdoAkkerman replied:
Hdiv and Hcurl are available now because of merge of #3999

dylan-copeland commented on miniapps/nurbs/nurbs_int.cpp:
```diff
@@ -0,0 +1,164 @@
+//
+// Description:  This example code illustrates usage of mixed finite element
+//               spaces, with three variants:
+//               0) H(div)
```
H1

IdoAkkerman replied:
Corrected.

IdoAkkerman commented on tests/unit/mesh/test_nurbs.cpp:
```diff
@@ -10,6 +10,7 @@
 // CONTRIBUTING.md for details.
 
 #include "mfem.hpp"
+#include <fstream> //TBD!!!
```
Removed.

IdoAkkerman commented on tests/unit/mesh/test_nurbs.cpp:
```diff
@@ -84,3 +85,112 @@ TEST_CASE("NURBS refinement and coarsening by spacing formulas", "[NURBS]")
+   mfem::out<<"knotvector : ";
+   kv.Print(mfem::out);
+
+   const int samples = 101;
```
Changed.

IdoAkkerman commented on tests/unit/mesh/test_nurbs.cpp:
```diff
@@ -84,3 +85,112 @@ TEST_CASE("NURBS refinement and coarsening by spacing formulas", "[NURBS]")
+   kv.Print(mfem::out);
+
+   const int samples = 101;
+   real_t u,xi,un;
```
Done.

IdoAkkerman commented on mesh/nurbs.hpp:
```diff
@@ -65,6 +71,9 @@ class KnotVector
        order @a order and number of control points @a NCP. */
    KnotVector(int order, int NCP);
 
+   /** @brief Create a KnotVector order @a order and knots @a knot. */
```
Corrected.

IdoAkkerman commented on mesh/nurbs.hpp:
```diff
@@ -94,13 +103,45 @@ class KnotVector
-   int findKnotSpan(real_t u) const;
+   MFEM_DEPRECATED int findKnotSpan(real_t u) const;  // Use GetSpan instead
+
+   /** Gives the @ i average knot location.
```
Done.

IdoAkkerman commented on mesh/nurbs.hpp:
```diff
@@ -177,6 +222,18 @@ class KnotVector
+       number of samples of the shape functions per element.*/
+   void PrintFunction(std::ostream &os, const Vector &a, int samples=11) const;
+
+   /** Prints the @a i-th functionand its first and second
```
Corrected.

IdoAkkerman commented on mesh/nurbs.cpp:
```diff
@@ -45,6 +45,54 @@ KnotVector::KnotVector(int order, int NCP)
    knot = -1.;
 }
 
+KnotVector::KnotVector(int order, Vector k)
```
Changed to `const Vector &k`. Also changed in hpp file.

IdoAkkerman commented on mesh/nurbs.cpp:
```diff
@@ -45,6 +45,54 @@ KnotVector::KnotVector(int order, int NCP)
+   if (repeated)
+   {
+      knot = k;
+      NumOfControlPoints = size - Order - 1;
```
Agree, removed.

IdoAkkerman commented on mesh/nurbs.hpp:
```diff
@@ -94,13 +103,45 @@ class KnotVector
+   /// Return the index of the knot span containing parameter @a u.
+   int GetSpan(real_t u) const;
+
+   /** @brief Return the reference coordinate in [0,1] for knot
```
Changed.

IdoAkkerman commented on mesh/nurbs.hpp:
```diff
@@ -94,13 +103,45 @@ class KnotVector
+
+   /** @brief Return the reference coordinate in [0,1] for knot
+       for the element beginning at knot @a ni. */
+   real_t GetIp(real_t u, int ni) const
```
Done.

IdoAkkerman commented on mesh/nurbs.hpp:
```diff
@@ -177,6 +222,18 @@ class KnotVector
        number of samples of the shape functions per element.*/
    void PrintFunctions(std::ostream &os, int samples=11) const;
 
+   /** Prints the function, parametrized by @a a, and its first and second
```
Changed to `Prints the function with basis function coefficient @a a,`

IdoAkkerman commented on mesh/nurbs.hpp:
```diff
@@ -53,6 +53,12 @@ class KnotVector
    /// Number of elements, defined by distinct knots.
    int NumOfElements;
 
+   // Stores the demko points
```
Will try to do this later today.

IdoAkkerman replied:
Done.

IdoAkkerman commented on mesh/nurbs.hpp:
```diff
@@ -94,13 +103,45 @@ class KnotVector
+   real_t GetIp(real_t u, int ni) const
+   { return (u-knot(ni))/(knot(ni+1)-knot(ni)); };
+
+   /** @brief Return the knot for element reference coordinate @a xi
```
Thanks for the   suggestion. `GetKnot` did not sit well with me either. I have changed it to `GetParam`

IdoAkkerman replied:
In the comments related to Greville, Demko and Botella points, I refer to knot location. Perhaps that is a better option as param or parameter is quit a generic name. This is also basically the same as the DEPRECATED function, except for the capital. I probably over did the refactoring...

IdoAkkerman commented:
@dylan-copeland thank you for the thorough and detailed review. There are still 3 things open on my end:

- Adding literature of Demko, Botella end Greville points
- Clarify the alternatives for the DEPRECATED `GetMaxima` and `GetInterpolant` routines.
- The 3 comments related to `GetSpan`.

I will try to address this points later today.
The other comments have been addressed.

IdoAkkerman commented on mesh/nurbs.hpp:
```diff
@@ -128,13 +169,17 @@ class KnotVector
        knot span @a xi, and the coordinate of the maximum in parameter space
        @a u. */
-   void FindMaxima(Array<int> &ks, Vector &xi, Vector &u) const;
+   MFEM_DEPRECATED void FindMaxima(Array<int> &ks, Vector &xi, Vector &u) const;
```
Added to the comments.

IdoAkkerman commented on mesh/nurbs.cpp:
```diff
@@ -60,6 +108,156 @@ KnotVector &KnotVector::operator=(const KnotVector &kv)
+
+real_t KnotVector::GetBotella(int i) const
+{
+   const int itermax = 10;
```
Done.

IdoAkkerman commented on mesh/nurbs.cpp:
```diff
@@ -45,6 +45,54 @@ KnotVector::KnotVector(int order, int NCP)
+   Order = order;
+
+   bool repeated = true;
+   int size = k.Size();
```
Done.

IdoAkkerman commented on mesh/nurbs.cpp:
```diff
@@ -60,6 +108,156 @@ KnotVector &KnotVector::operator=(const KnotVector &kv)
+{
+   int low, mid, high;
+
+   if (u == knot(NumOfControlPoints+Order))
```
Added.

IdoAkkerman commented on mesh/nurbs.cpp:
```diff
@@ -60,6 +108,156 @@ KnotVector &KnotVector::operator=(const KnotVector &kv)
+      low = Order;
+      high = NumOfControlPoints + 1;
+      mid = (low + high)/2;
+      while ( (u < knot(mid)) || (u > knot(mid+1)) )
```
If `u=1` will results in `if (u == knot(NumOfControlPoints+Order))` to be used, giving `mid = NumOfControlPoints-1`
as result.
That being said, the routine was a bit questionable before I changed things and it is worth to improve the overall quality and robustness of the routine.

I changed the comparison to `>=` and it  initially did not change the validity of the unit test.
I have added your case to the test cases.
I also modified the test vector (with repeated knot left and right of the middle, and the middle as well),
this did trigger the issue!!
**Long story short, I adopted the suggested change.**


Note: I also changed the starting location, viz: `high = NumOfControlPoints ;`

IdoAkkerman commented:
The remaining comments have been addressed.

dylan-copeland commented on mesh/nurbs.hpp:
```diff
@@ -127,14 +190,26 @@ class KnotVector
-   void FindMaxima(Array<int> &ks, Vector &xi, Vector &u) const;
+       @a u.
+       The main purpose of this function is its use in FindInterpolant.
+       Use GetBotella instead for each shape function seperately, perhaps in
```
separately

IdoAkkerman replied:
Corrected.

dylan-copeland commented on examples/ex1.cpp:
```diff
@@ -143,7 +143,7 @@ int main(int argc, char *argv[])
    if (order > 0)
    {
-      fec = new H1_FECollection(order, dim);
+      fec = FECollection::NewH1(order, dim, mesh.NURBSext);
```
This is fine, but we might consider a member function in `Mesh`:
`bool isNURBS() const { return NURBSext; }`

IdoAkkerman replied:
Added this function to the mesh object.

dylan-copeland commented on fem/fe_coll.hpp:
```diff
@@ -174,6 +174,27 @@ class FiniteElementCollection
    */
    static FiniteElementCollection *New(const char *name);
 
+   /** Factory method: return a newly allocated H1 conforming FiniteElementCollection
```
Although I never use doxygen, I guess we should follow the convention of using `@brief`. Some periods in these comments would also help readability.

IdoAkkerman replied:
Added the `@ brief` 4-times.

dylan-copeland commented on fem/fe/fe_nurbs.hpp:
```diff
@@ -86,6 +86,10 @@ class NURBS1DFiniteElement : public ScalarFiniteElement,
                            DenseMatrix &dshape) const;
    virtual void CalcHessian (const IntegrationPoint &ip,
                              DenseMatrix &hessian) const;
+   virtual void Project(Coefficient &coeff,
```
In the derived classes, `override` should be used instead of `virtual`, but `virtual` appears to be used in many derived classes. It looks like this PR is following the general pattern in MFEM which should be corrected in a future PR. I suppose we can keep this for now.

IdoAkkerman replied:
Replaced the virtuals with override.

dylan-copeland commented on mesh/nurbs.hpp:
```diff
@@ -94,13 +103,67 @@ class KnotVector
    int GetNKS() const { return NumOfControlPoints - Order; }
 
-   /** @brief Return the parameter for element reference coordinate @a xi
+   /// Return whether knot @a u is in a given span @a ni.
```
`parameter @a u`?

IdoAkkerman replied:
Changed to knot location.

dylan-copeland commented on fem/gridfunc.cpp:
```diff
@@ -2393,29 +2393,39 @@ void GridFunction::ProjectCoefficient(Coefficient &coeff)
       }
       else
       {
-         // Define and assemble linear form
```
This replaces the global L2 projection with local interpolation, in the NURBS case. There is a comment for this function in gridfunc.hpp explaining that global L2 projection is used in the NURBS case. That comment should be updated. Should we also keep the global L2 projection in a new function, e.g. `ProjectCoefficientGlobalL2`, which could be called for any space (not just NURBS)?

dylan-copeland replied:
Should there be a parallel version in `ParGridFunction`, with `ProjectCoefficientGlobalL2` being a virtual function? It looks like in parallel, the `GridFunction:: ProjectCoefficientGlobalL2` will be called, which assembles serial systems on each MPI rank. This is an L2 projection on each MPI rank's subdomain.

dylan-copeland commented on fem/gridfunc.cpp:
```diff
@@ -2393,29 +2393,39 @@ void GridFunction::ProjectCoefficient(Coefficient &coeff)
-         cg.Mult(b,*this);
+         Array<int> vdofs;
+         Vector vals;
+         real_t signal = std::numeric_limits<real_t>::min();
```
`constexpr`

IdoAkkerman replied:
Added.

dylan-copeland commented on fem/pgridfunc.cpp:
```diff
@@ -566,7 +566,13 @@ void ParGridFunction::ProjectCoefficient(Coefficient &coeff)
 
    if (delta_c == NULL)
    {
+      (*this) = std::numeric_limits<real_t>::min();
```
Should this initialization be done in `GridFunction::ProjectCoefficient` instead, so that the non-parallel case is also initialized? I suppose it doesn't matter, if there are undefined DOFs only in parallel, and that must be true if we do not initialize in the base class `GridFunction`.

IdoAkkerman replied:
Line 569 forms a tandem with line 574. Therefor I like to fact that they are in close proximity.
This is a potential point of confusion and fragility.  I will document this better here and in the accompanying gridfunction memberfunction.

dylan-copeland commented on miniapps/nurbs/nurbs_mesh_info.cpp:
```diff
@@ -0,0 +1,166 @@
+// Sample runs:
+//    nurbs_mesh_info -m ../data/cude_nurbs.mesh -p 0 -r 2
+
+// Description:  This example code solves the time-dependent advection equation
```
This copied description is not right.

IdoAkkerman replied:
Added correct text.

dylan-copeland commented on miniapps/nurbs/nurbs_mesh_info.cpp:
```diff
@@ -0,0 +1,166 @@
+// Compile with: make nurbs_mesh_info
+//
+// Sample runs:
+//    nurbs_mesh_info -m ../data/cude_nurbs.mesh -p 0 -r 2
```
nurbs_mesh_info -m ../../data/cube-nurbs.mesh -r 2

IdoAkkerman replied:
Fixed.

dylan-copeland commented on miniapps/nurbs/CMakeLists.txt:
```diff
@@ -37,6 +37,10 @@ add_mfem_miniapp(nurbs_printfunc
   MAIN nurbs_printfunc.cpp
   LIBRARIES mfem)
 
+add_mfem_miniapp(nurbs_mesh_info
```
Also add to `makefile`? It doesn't build with `make`.

IdoAkkerman replied:
Done.

dylan-copeland commented on miniapps/nurbs/nurbs_int.cpp:
```diff
@@ -0,0 +1,164 @@
+      }
+   }
+
+   // 5. Define a finite element space on the mesh. Here we use Nedelec or
```
H1, ND, or RT

IdoAkkerman replied:
Changed to:

Here we use H_1, H(Div) or H(Curl) finite elements of the specified order.

IdoAkkerman replied:
Also in nurbs_ex24

dylan-copeland commented on fem/gridfunc.hpp:
```diff
@@ -387,10 +387,21 @@ class GridFunction : public Vector
+       NOTE: For parallel simulations with NURBS elements some dofs might
+       not be defined, if the evaluation point does not reside on this rank.
+       If that is the case it is defined on another rank, and the issue is
+       rectified with the appropriate communcation, see in ParGridFunction.
```
`communication`

IdoAkkerman replied:
Fixed (2x)

dylan-copeland commented on fem/gridfunc.hpp:
```diff
@@ -399,10 +410,20 @@ class GridFunction : public Vector
+       NOTE: For parallel simulations with NURBS elements some dofs might
+       not be defined, if the evaluation point does not reside on this rank.
+       If that is the case it is defined on another rank, and the issue is
+       rectified with the appropriate communcation, see in ParGridFunction.*/
```
`communication`

IdoAkkerman replied:
Fixed (2x)

dylan-copeland commented on fem/gridfunc.cpp:
```diff
@@ -2428,6 +2444,34 @@ void GridFunction::ProjectCoefficient(Coefficient &coeff)
+   cg.SetOperator(A);
+   cg.SetPreconditioner(prec);
+   cg.SetRelTol(1e-12);
+   cg.SetMaxIter(1000);
```
`1000` -> `iter`, also `rtol`

IdoAkkerman replied:
Indeed.

dylan-copeland commented on fem/gridfunc.cpp:
```diff
@@ -2476,37 +2517,72 @@ void GridFunction::ProjectCoefficient(VectorCoefficient &vcoeff)
+   CGSolver cg;
+   cg.SetOperator(A);
+   cg.SetPreconditioner(prec);
+   cg.SetRelTol(1e-12);
```
`rtol, iter`

IdoAkkerman replied:
I should check my sown stuff more/better...

IdoAkkerman commented on fem/fe_coll.hpp:
```diff
@@ -174,6 +174,27 @@ class FiniteElementCollection
+       Caller gets ownership. */
+   static FiniteElementCollection *NewL2(int order, int dim, bool NURBS = false);
+
+   /** F@brief actory method: return a newly allocated HDiv conforming FiniteElementCollection.
```
Fixed

dylan-copeland commented on fem/pgridfunc.cpp:
```diff
@@ -582,6 +588,71 @@ void ParGridFunction::ProjectCoefficient(Coefficient &coeff)
+   b.Assemble();
+
+   // Set solver and preconditioner
+   SparseMatrix A(a.SpMat());
```
This should be like `ex1p`, rather than using the serial `SparseMatrix`.

IdoAkkerman replied:
Fixed it.

IdoAkkerman commented on fem/gridfunc.cpp:
```diff
@@ -2476,35 +2517,76 @@ void GridFunction::ProjectCoefficient(VectorCoefficient &vcoeff)
+   // Define and assemble linear form
+   LinearForm b(fes);
+   BilinearForm a(fes);
+   if (fes->FEColl()->GetRangeType(3)  == mfem::FiniteElement::VECTOR)
```
As the `GetRange` is asked to the `FiniteElementCollection` as a whole and not a particular `Element` a dimension needs to be provided. I chose 3, but this could also be 2, that is immaterial from this perspective.
Could this create issues on the other side? I think we should cross that bridge when we are there...
For now I will add a comment that this dimension is chosen arbitrarily.

dylan-copeland replied:
Thanks for documenting. I thought this dimension should come from the `Mesh::Dimension()`, but I think you are right that 2 or 3 should work.

dylan-copeland commented on miniapps/nurbs/nurbs_ex24.cpp:
```diff
@@ -126,8 +126,8 @@ int main(int argc, char *argv[])
 
-   // 5. Define a finite element space on the mesh. Here we use Nedelec or
-   //    Raviart-Thomas finite elements of the specified order.
+   // 5. Define a finite element space on the mesh. Here we use H_1, H(Div) or
```
For consistency, `H^1` is used above on line 26.

IdoAkkerman replied:
Fixed

dylan-copeland commented on miniapps/nurbs/nurbs_int.cpp:
```diff
@@ -0,0 +1,164 @@
+   int order = 1;
+   bool NURBS = true;
+   int prob = 0;
+   bool static_cond = false;
```
Unused: `static_cond`, `pa`, `visualization`.

IdoAkkerman replied:
The `nurbs_int` was meanly a debug vehicle. There is no real need for the app, so I decided to remove it.

dylan-copeland commented on miniapps/nurbs/nurbs_int.cpp:
```diff
@@ -0,0 +1,164 @@
+//                                MFEM Example interpolation -- modified for NURBS FE
+//
+// Compile with: make nurbs_int
```
This file is not compiled with `make` in the `nurbs` directory. I guess it needs to be added to `makefile`.

IdoAkkerman replied:
app is removed

dylan-copeland commented on miniapps/nurbs/nurbs_int.cpp:
```diff
@@ -0,0 +1,164 @@
+//
+// Compile with: make nurbs_int
+//
+// Sample runs:  nurbs_int -m ../../data/pipe-nurbs-2d.mesh -o 2
```
There is no `-m` option.

IdoAkkerman replied:
app is removed

IdoAkkerman commented on miniapps/nurbs/nurbs_ex1.cpp:
```diff
@@ -409,7 +423,8 @@ int main(int argc, char *argv[])
    GridFunction x(fespace);
-   x = 0.0;
+   FunctionCoefficient  sol_cf(sol);
+   x.ProjectCoefficient(sol_cf);
```
I added a command line option, with the default being homogeneous BC(= old behavior)

tzanio commented:
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

justinlaughlin commented on mesh/nurbs.cpp:
```diff
@@ -45,6 +45,53 @@ KnotVector::KnotVector(int order, int NCP)
    knot = -1.;
 }
 
```
I like this constructor; the `repeated` check is a nice safeguard but it also means this will often initialize a `KnotVector` with a vector different from what was provided. IMO adding either a warning or even just a small comment in the header (I think that is sufficient) would be helpful.

justinlaughlin commented on mesh/nurbs.cpp:
```diff
@@ -60,6 +107,165 @@ KnotVector &KnotVector::operator=(const KnotVector &kv)
+   return sum/real_t(Order);
+}
+
+real_t KnotVector::GetBotella(int i) const
```
Do you think it might be worth it to cache the results of `GetBotella`? I noticed `KnotVector` now stores Demko points ([nurbs.hpp](https://github.com/mfem/mfem/pull/4326/files#diff-c75875c788ff181d814edcb659958b967abf8742fcfcc35e64a10766d87f397cR57)) but throughout the code it seems like Botella points are used quite frequently (maybe more frequently than Demko points?)

justinlaughlin commented on mesh/nurbs.cpp:
```diff
@@ -60,6 +107,165 @@ KnotVector &KnotVector::operator=(const KnotVector &kv)
+   return u;
+}
+
+void KnotVector::ComputeDemko() const
```
I was testing out Greville/Botella/Demko points for various knotvectors. One thing I discovered is that it seems like if there is an internal knot with multiplicity `>= p+1` `ComputeDemko()` will fail, while the other routines wont. As an example 
```cpp
KnotVector kv(2, Vector({0.0,1.0,1.0,1.0,2.0}));
Vector greville(kv.GetNCP());
for (int i = 0; i < kv.GetNCP(); i++)
{
   greville[i] = kv.GetGreville(i);
}
mfem::out<<"Greville points : "; greville.Print(mfem::out, 32);


Vector botella(kv.GetNCP());
for (int i = 0; i < kv.GetNCP(); i++)
{
   botella[i] = kv.GetBotella(i);
}
mfem::out<<"Botella  points : "; botella.Print(mfem::out, 32);


Vector demko(kv.GetNCP());
for (int i = 0; i < kv.GetNCP(); i++)
{
   demko[i] = kv.GetDemko(i);
}
mfem::out<<"Demko    points : "; demko.Print(mfem::out, 32);
```

returns 
```cpp
Greville points : 0 0.5 1 1 1.5 2
Botella  points : 0 0.5 1 1 1.5 2

DenseMatrix::Invert() : singular matrix
```
B-splines for reference
<img width="594" alt="Screenshot 2024-10-14 at 10 07 42 PM" src="https://github.com/user-attachments/assets/b33b8f59-29da-4d05-8ca1-f593a5843d39">

Looks like its coming from the `GetInterpolant()` routine. Maybe this is expected and there is nothing to fix; in that case a note would be useful.

btw, I did a brute-force check on the Greville/Botella points for many different knotvectors and the values looked correct 👍

justinlaughlin reviewed (commented):
I left a few comments but otherwise this looks good to me! There are a lot of improvements here, thank you.

justinlaughlin commented on mesh/nurbs.hpp:
```diff
@@ -65,6 +71,9 @@ class KnotVector
        order @a order and number of control points @a NCP. */
    KnotVector(int order, int NCP);
 
+   /** @brief Create a KnotVector with order @a order and knots @a knot. */
```
```suggestion
   /** @brief Create an open-ended KnotVector with order @a order and knots @a knot;
   internal knots have a multiplicity of 1. */
```

justinlaughlin replied:
4733519

justinlaughlin commented:
@IdoAkkerman, @JacobLotz - this code would be very nice to merge in, several other branches could make use of features here e.g. Greville points. If you have a chance could you merge `master` into here and then we can move this to `next`? Thank you

justinlaughlin commented on mesh/nurbs.cpp:
```diff
@@ -100,6 +147,165 @@ KnotVector &KnotVector::operator=(const KnotVector &kv)
+
+void KnotVector::ComputeDemko() const
+{
+   if (demko.Size() == GetNCP()) { return; }
```
Placing this check in `GetDemko` accomplishes the same thing but makes it possible to recompute points if necessary.

(`KnotVector` is not immutable e.g. `Flip`)

tzanio commented:
Merged in `next` for testing...

v-dobrev commented on examples/ex1.cpp:
```diff
@@ -143,7 +143,7 @@ int main(int argc, char *argv[])
    if (order > 0)
    {
-      fec = new H1_FECollection(order, dim);
+      fec = FECollection::NewH1(order, dim, mesh.IsNURBS());
```
I'm not sure we want this in Example 1.

tzanio replied:
You are right, we don't.

justinlaughlin replied:
Reverted in b02c99b9e

tzanio commented:
Re-merged in `next` for testing...

camierjs commented:
There were some changes to bring from `doftrans-three-arg`, done in 2bef31b and  4a46614.
We still need the `doftrans.GetDofTransformation()` `if` statement in this branch for the tests to pass, but we should be able to remove it when re-merging in `next`.

v-dobrev commented:
Re-merged in `next` for testing...

v-dobrev commented:
@justinlaughlin, I now see that the change you made is in `ex24` and not `nurbs_ex24` (as I assumed without looking carefully). Why change `ex24` which shows no diffs in the night tests and not `nurbs_ex24`? I guess I assumed you changed `nurbs_ex24` due to your comment:
> Changing `ProjectCoefficient` to `ProjectCoefficientGlobalL2` throughout `nurbs_ex24` recovers the original results.

Another thing that sees inconsistent: why change `ex24` but not `ex24p`? Also, the changes in `ex24` affect non-NURBS spaces which is not what we want, I think.

Ultimately, even if we make `ProjectCoefficientGlobalL2` the default for some cases, we still want to ensure that `ProjectCoefficient` works reasonably well for NURBS -- currently, this does not seem to be the case based on the screenshots you posted above.

v-dobrev commented on fem/gridfunc.cpp:
```diff
@@ -2410,30 +2447,73 @@ void GridFunction::ProjectCoefficient(VectorCoefficient &vcoeff)
 
-      // Define and assemble bilinear form
-      BilinearForm a(fes);
+   // Dimension argument to GetRangeType is arbitrary to be 3, could also be 2.
+   if ((fes->FEColl()->GetRangeType(3) == mfem::FiniteElement::VECTOR) ||
+       (fes->FEColl()->GetRangeType(2) == mfem::FiniteElement::VECTOR))
```
This check should not be randomly using 3 or 2 ... it should use precisely the dimension of the mesh.

Alternatively, it can be replaced by this check:
```suggestion
   if (fes->GetTypicalFE()->GetRangeType() == mfem::FiniteElement::VECTOR)
```

v-dobrev commented:
Whatever the reason for the issue we see in https://github.com/mfem/mfem/pull/4326#issuecomment-2982327754 -- it needs to be resolved before merging this PR. Replacing `ProjectCoefficient` with `ProjectCoefficientGlobalL2` in `nurbs_ex24` is masking the issue, so we need to have a way to reproduce it. One option could be to add a command-line option that allows us to switch between `ProjectCoefficientGlobalL2` and `ProjectCoefficient`.

v-dobrev commented:
Hi @IdoAkkerman,

Do you know what is the expected convergence order of the new NURBS nodal interpolation, as defined by `ProjectCoefficient`? Is it expected to be optimal?

We observe results (when using the new NURBS nodal interpolation) indicating that there is either a bug in the implementation, or that the interpolation is not very good, at least near the boundaries, see https://github.com/mfem/mfem/pull/4326#issuecomment-2982327754 and https://github.com/mfem/mfem/pull/4326#issuecomment-2989092191.

In the current state of this branch, we added an option to `nurbs_ex24.cpp` to switch between the global L2 projection, `-gp`/`--global-projection` (default), and the new local interpolation, `-no-gp`/`--no-global-projection`. We made the former the default (for now) because the errors we get with the latter are much larger.

v-dobrev commented:
Re-merged in `next` for testing...

tzanio commented:
As agreed with @justinlaughlin and @dylan-copeland, portions of this branch will be added to separate PRs.

IdoAkkerman commented:
@tzanio 

Should I split the PR in three separate PRs?

I was thinking:

1) The "Nodal projection part"

2) The fespace initialization part

3) The shape function selection utility functions

Or did you have an other split in mind?

IdoAkkerman commented:
I implemented the "Bezier Projection" and the issue at the boundaries is gone.
I think the name "Bezier Projection" is a bit of a misnomer, as I think it would work on a large class of elements (all perhaps?). The algorithm consists of two steps:
1) Project the solution on the local basis by means of an L2 projection.
2) DOFs that are shared between element take a weighted average.

The papers splits step 1 in two substeps involving the Bezier shape functions. I do not see any purpose for this, other than conditioning perhaps?

IdoAkkerman commented:
Conditioning is an issue for higher order NURBS  unfortunately...

justinlaughlin commented:
> I implemented the "Bezier Projection" and the issue at the boundaries is gone.

That's great to hear 😄. What is the issue with conditioning you are seeing with higher order?

> [direct injection in the coefficient of the value at the evaluated point...]

Am I understanding correctly that the value at Botella points is being used directly *as the coefficients*? I would think that after obtaining values at Botella points we should still be solving a system for the coefficients (using e.g. `GetInterpolant`); or alternatively performing a local approximation such as in the linked paper. This should still be much faster than the previous "global l2" projection since each system of coefficients is just for a single patch + dimension. #4841 added a banded solver to `FindInterpolant` (which is deprecated here).

However if you are getting better results with the "Bezier Projection" than perhaps this is a moot point 😀.

> Should I split the PR in three separate PRs?

I think your suggested split of the PR makes sense.

IdoAkkerman commented:
When looking at the shapefunctions of higher order NURBS away from the boundary, you will see a very long tail (A bit like a gauss curve). In these tails the shapefunctions have very low values and look very similar. In the paper they side step this issue somewhat by first projecting on the Bernstein polynomials and from there "reconstruct"(= also project) the NURBS coefficients. 

Direct projection worked until p=3, using the Bernstein polynomials  (H1 positive) worked until p=6  (Note that the paper only presents up to p=5). By using Jacobi polymials (L2) I was able to go to p=9.

So the conditioning can be circumvented by having a two step projection. Where Jacobi polynomials work better then the bernstein polynomials.

IdoAkkerman commented:
Correct Botella points "my" projection basically  assumes the NURBS are interpolatory (=kronecker property) at the Botella 
points. This is off course not true, but the assumption is that under mesh refinement these errors made will converge to zero. 

I think the direct injection is a necessary feature (evil?) if you want to strictly satisfy the Min/Max property.
When solving the system such as to guarantee the solution to be interpolatory  at specifics points you would loss this property, in this case I think you would be better of using the Demko points.

Side note: Demko points are related to the function space spanned, while the Botella points are  directly related to the basis used. So using a different basis to span the same space will result in identical Demko points but different Botella points.


I see how using finding an interpolant can be done patch wise, assuming patches use the same points on shared faces/edges. However, I do not see how the different dimensions could be split easily..
Moreover solving Patch wise problems in parallel seems like an unnecessary complication.

IdoAkkerman commented:
I will work further on the LocalL2 projection. I will also implement a vectorFE and array of standerd FE version for the VectorCoefficient case, as well the parallel versions of those. I will use nurbs_ex24, nurbs_ex24p, nurbs_ex10 and nurbs_ex10p for debugging and testing.

justinlaughlin commented:
> @justinlaughlin This PR also touched some interpolation routines, due to the Demko/Botella point addition.
I tried the manually merge the confilcts but was not aware these routines had changed, so I accidentally brook it....
I tried to fix it.

Is this related to `FindInterpolant` vs `GetInterpolant`? @dylan-copeland added a banded solver in #4841 to the former which I believe is deprecated in this PR

IdoAkkerman commented:
@justinlaughlin Yes I deprecated it, but that was before the @dylan-copeland changes.

@dylan-copeland Could you perhaps check if I did not break your stuff...
I think the Demko Points would actually be a beter choice for nodal interpolation.

IdoAkkerman commented:
This PR requires #4942

IdoAkkerman commented:
I will be AFK till August 1.

IdoAkkerman commented:
Would it be okay to add a flag to the following routines:

```
virtual void ProjectCoefficient(Coefficient &coeff);
virtual void ProjectCoefficient(VectorCoefficient &vcoeff);
```
such that it becomes:
```
 enum ProjType
 {
      DEFAULT,
      ELEMENT,  
      ELEMENTL2,
      GLOBALL2
 };

virtual void ProjectCoefficient(Coefficient &coeff, ProjType type  = ProjType::DEFAULT);
virtual void ProjectCoefficient(VectorCoefficient &vcoeff, ProjType type  = ProjType::DEFAULT);
```

where `DEFAULT` is `ELEMENT` for NonNURBS elements and  ELEMENTL2 for NURBS elements

justinlaughlin commented:
> Would it be okay to add a flag to the following routines:
> 
> ```
> virtual void ProjectCoefficient(Coefficient &coeff);
> virtual void ProjectCoefficient(VectorCoefficient &vcoeff);
> ```
> 
> such that it becomes:
> 
> ```
>  enum ProjType
>  {
>       DEFAULT,
>       ELEMENT,  
>       ELEMENTL2,
>       GLOBALL2
>  };
> 
> virtual void ProjectCoefficient(Coefficient &coeff, ProjType type  = ProjType::DEFAULT);
> virtual void ProjectCoefficient(VectorCoefficient &vcoeff, ProjType type  = ProjType::DEFAULT);
> ```
> 
> where `DEFAULT` is `ELEMENT` for NonNURBS elements and ELEMENTL2 for NURBS elements

I like this solution 👍

tzanio commented:
Let me know if/when this is `ready-for-review` guys

dylan-copeland commented on fem/coefficient.hpp:
```diff
@@ -1858,6 +1858,85 @@ class TraceCoefficient : public Coefficient
+   VectorCoefficient *a;// = nullptr;
+
+   mutable Vector va;
+   int i;
```
A longer, more descriptive name like `component` may be safer.

IdoAkkerman replied:
Changed

dylan-copeland commented on fem/coefficient.hpp:
```diff
@@ -1858,6 +1858,85 @@ class TraceCoefficient : public Coefficient
+   MatrixCoefficient *a = nullptr;
+
+   mutable DenseMatrix ma;
+   int i,j;
```
Could these be renamed more descriptively, e.g. `row_index` and `column_index`?

dylan-copeland commented on fem/coefficient.hpp:
```diff
@@ -1858,6 +1858,85 @@ class TraceCoefficient : public Coefficient
+class VectorComponentCoefficient : public Coefficient
+{
+private:
+   VectorCoefficient *a;// = nullptr;
```
It seems to work with `= nullptr;`

dylan-copeland commented on miniapps/nurbs/CMakeLists.txt:
```diff
@@ -50,6 +50,11 @@ add_mfem_miniapp(nurbs_ex5
   add_dependencies(nurbs_ex5 copy_miniapps_nurbs_data)
 
+add_mfem_miniapp(nurbs_ex10
+  MAIN nurbs_ex10.cpp
```
Also add to `makefile` (this example is not built with `make`). Similarly for `nurbs_ex10p.cpp`.

IdoAkkerman replied:
Done. Also updated .gitignore

dylan-copeland commented on miniapps/nurbs/nurbs_ex10.cpp:
```diff
@@ -0,0 +1,623 @@
+// Compile with: make nurbs_ex10
+//
+// Sample runs:
+//    nurbs_ex10 -m ../../data/beam-quad-nurbs.mesh -s 23 -r 2 -o 2 -dt 3
```
The first sample run has a Newton convergence failure.

dylan-copeland replied:
This first sample run is still failing.

justinlaughlin replied:
I think it makes sense to remove non-NURBS from the NURBS miniapps.

IdoAkkerman replied:
I agree, having the capability has its -- limited -- value in debugging, but the default meshes should be NURBS meshes, and tests should only use NURBS meshes.

I think changing the defaults will fix the tests.

dylan-copeland commented on fem/gridfunc.cpp:
```diff
@@ -2390,49 +2582,316 @@ void GridFunction::ProjectCoefficient(
+   {
+      Array<int> vdofs(fes->GetNDofs());
+      Vector x, Va;
+      VectorComponentCoefficient coeff(vcoeff,0);
```
In the scalar range case, the first component of the `VectorCoefficient` is taken. Why is this policy needed or used? If it is really needed, it should be documented.

IdoAkkerman replied:
Add comment to clarify

dylan-copeland commented on fem/coefficient.hpp:
```diff
@@ -1858,6 +1858,85 @@ class TraceCoefficient : public Coefficient
+};
+
+/// Scalar coefficient defined as component of a matrix coefficient
+class MatrixComponentCoefficient : public Coefficient
```
Is this class used? Everything builds if it is commented out.

dylan-copeland commented on .gitignore:
```diff
@@ -319,6 +319,8 @@ miniapps/nurbs/nurbs_naca_cmesh
 miniapps/nurbs/naca-cmesh.mesh
 miniapps/nurbs/glvis_naca-cmesh.mesh
 miniapps/nurbs/Naca_cmesh
+miniapps/nurbs/nurbs_mesh_info
```
Add `nurbs_ex10` and `nurbs_ex10p`

IdoAkkerman commented on mesh/nurbs.hpp:
```diff
@@ -103,13 +125,73 @@ class KnotVector
-   int findKnotSpan(real_t u) const;
+   MFEM_DEPRECATED int findKnotSpan(real_t u) const;  // Use GetSpan instead
+
+   /** Gives the @a i average knot location. Average is taken over @a Order
```
`i` is neither the index indicating a control point or basis function. It is the index in the vector or greville points...
Explaining how it is obtained (and how it is connected with the knotvector) does not seem relevant for me IMHO.

Sure you can associate Greville points with basis functions, but mentioning this here seems more confusing then helpfull, IMHO.

IdoAkkerman commented on mesh/nurbs.hpp:
```diff
@@ -103,13 +125,73 @@ class KnotVector
+   /** @brief Return the parameter for element reference coordinate @a xi
+       in [0,1], for the element beginning at knot @a ni. */
+   MFEM_DEPRECATED real_t getKnotLocation(real_t xi, int ni) const
+   { return (xi*knot(ni+1) + (1. - xi)*knot(ni)); } // Use GetKnot instead
```
This is an old function. Therefor the MFEM_DEPRICATED statement.
Changing the name would defeat the purpose...

dylan-copeland replied:
I mean the comment `Use GetKnot instead` should be `Use GetKnotLocation instead`.

IdoAkkerman commented on mesh/nurbs.cpp:
```diff
@@ -724,6 +1065,117 @@ void KnotVector::FindInterpolant(Array<Vector*> &x, bool reuse_inverse)
+{
+   int ncp = GetNCP();
+
+   // Initialize matrix
```
Duplication is because `GetInterpolant` is meant as a replacement for  `FindInterpolant`, hence that the later is now made DEPRECATED.

dylan-copeland replied:
Right, the duplication is temporary, but I was thinking even the temporary duplication could be avoided by refactoring. I guess it's fine to leave it as it is.

IdoAkkerman commented on .gitignore:
```diff
@@ -319,6 +319,8 @@ miniapps/nurbs/nurbs_naca_cmesh
 miniapps/nurbs/glvis_naca-cmesh.mesh
 miniapps/nurbs/Naca_cmesh
+miniapps/nurbs/nurbs_mesh_info
+miniapps/nurbs/k[0-1]_*.dat
```
done

IdoAkkerman commented on mesh/nurbs.hpp:
```diff
@@ -103,13 +125,73 @@ class KnotVector
+       Reverts to the Greville point if knot is repeated @a Order +1 times.
+       For background see:
+
+       Botella, Olivier, and Karim Shariff.
```
done

IdoAkkerman commented on mesh/nurbs.hpp:
```diff
@@ -63,6 +78,13 @@ class KnotVector
+       If @a k has repeated knots, this constructor will copy the knots
+       exactly as they are provided. Otherwise, end knots will be repeated
+       (order + 1) times (C^{-1}) while internal knots will retain a
+       multiplicity of 1. */
```
Clarified.

IdoAkkerman commented on mesh/nurbs.cpp:
```diff
@@ -108,6 +155,241 @@ KnotVector &KnotVector::operator=(const KnotVector &kv)
+   Vector x(GetNCP());
+   for ( int i = 0; i <x.Size(); i++)
+   {
+      x[i] = std::pow(-1.0, i);
```
done.

IdoAkkerman commented on mesh/nurbs.hpp:
```diff
@@ -103,13 +125,73 @@ class KnotVector
+
+       Points are found using Remez iteration:
+        - Find interpolant, given by a, through given points, given by Demko
+        - Find extrema of this polynom and update Demko points
```
done

IdoAkkerman commented on mesh/nurbs.cpp:
```diff
@@ -108,6 +155,241 @@ KnotVector &KnotVector::operator=(const KnotVector &kv)
+            }
+         }
+         // Check convergence
+         /*if ((itermax2 > 1) &&
```
done

IdoAkkerman commented on fem/gridfunc.hpp:
```diff
@@ -390,12 +394,51 @@ class GridFunction : public Vector
+       Note 2: For NURBS elements the ELEMENT projection gives results without
+       over and undershoots. However, the gradient near the boundary does not
+       converge.*/
+   enum ProjType { DEFAULT, ELEMENT, GLOBALL2, ELEMENTL2 };
```
done


When used it becomes:
```
GridFunction::ProjType::ELEMENT
```
It is a bit long. Perhaps that is no problem, or should we move the enum class outside of the GridFunction definition, such that would become:
``
ProjType::ELEMENT
```

IdoAkkerman replied:
I moved it outside the class. I also added the underscore.
I also changed `ProjType` -->  `ProjectType`

IdoAkkerman commented:
@dylan-copeland Thanks for the comments. I have update the PR, using most of the pointers provided.
Please see the response for the where and why when I did not follow-up on the provided pointer.

justinlaughlin reviewed (approved):
Sorry, forgot to click "Submit Review". Thank you for all the changes Ido.

v-dobrev commented on CHANGELOG.this:
Is this meant to be in `CHANGELOG`? I don't think we want a separate file like this one.

justinlaughlin commented:
Very glad to see this merged! 😄 Thanks @IdoAkkerman !
